### PR TITLE
Move in Bugzilla integration

### PR DESCRIPTION
**Pull request by CyberShadow**
Here's a first (untested) draft for the Bugzilla integration.

Any thoughts on how to test this?

Some things left to do:

- What if the issue is already closed? Need to check if the Bugzilla API errors out. We should post a comment regardless.
- Check the comment formatting and see if we want to put any more data there, such as the commit message which contains the reference to the bug in question. Will need to split up closing per-issue instead of in bulk.
- Needs the Bugzilla auth in environment as for the other services.

### CODE REVIEW

**Comment by wilzbach** [comment]
> Here's a first (untested) draft for the Bugzilla integration.

Awesome!

> Any thoughts on how to test this?

I added a test to your PR.

tl;dr:
- there's `/data/payloads` which the testsuite will try to check if a payload exists and return that
- you can overwrite this with manually handling the request with Vibe.d:

```d
"/github/repos/dlang/dmd/pulls/6327/merge",
(scope HTTPServerRequest req, scope HTTPServerResponse res){  ... }
```

- if an URL that's not in the API expectations has been called, the testsuite errors, otherwise it will be removed (that's why some URLs appear twice, e.g. GET and POST)

For this to work, the bugzilla URL needs to point to the fake test server, but `bugzillaURL` already does that.
To run the test suite, use sth. like:

```
dub test --override-config vibe-d:tls/openssl-1.1 -- --single --trace
```

Or just `dub test --override-config vibe-d:tls/openssl-1.1 -- -s bugzilla.after-merge-close-issue-bugzilla`.

> * What if the issue is already closed? Need to check if the Bugzilla API errors out. We should post a comment regardless.

Hmm there's also the thing that dlang-bot might have already closed the issue because it was merged on `stable`, so we probably should check for existing comments from dlang-bot to avoid duplicate posts.

> * Check the comment formatting and see if we want to put any more data there, such as the commit message which contains the reference to the bug in question. Will need to split up closing per-issue instead of in bulk.

That's imho secondary. I think for now it's just important that we get this integration working.

> * Needs the Bugzilla auth in environment as for the other services.

Ok. I created an account for the dlang-bot on Bugzilla and added the environment variables to the Heroku instance. You should have access to that instance too in case you ever want to e.g. check the logs.

**Comment by wilzbach** [comment]
> I added a test to your PR.

FYI: The test is a bit fake as I didn't know what Bugzilla is supposed to reply, but I hope it shows how the testsuite here works.

**Comment by CyberShadow** [comment]
That's great, thanks! I'll tinker with this some more tomorrow.

File: test/bugzilla.d [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+         },
+        "/github/repos/dlang/phobos/issues/4963/comments",
+        "/github/repos/dlang/phobos/issues/4963/labels",
+        (scope HTTPServerRequest req, scope HTTPServerResponse res){
+            res.writeJsonBody(Json.emptyArray);
+        },
+        "/github/repos/dlang/phobos/issues/4963/labels",
+        (scope HTTPServerRequest req, scope HTTPServerResponse res){
+            res.writeVoidBody;
+        },
```
Comment (wilzbach) [inline test/bugzilla.d]: We should probably don't rewrite the comment and update the labels when a PR gets merged (that's why I had to add these three expectations), but this can be done in another PR.

File: source/dlangbot/bugzilla.d [hunk 2]
```diff
@@ -79,3 +81,43 @@ Issue[] getDescriptions(R)(R issueRefs)
...
+Json apiCall(string method, Json[string] params)
+{
+    import std.net.curl;
+    import vibe.stream.operations : readAllUTF8;
+    import dlangbot.utils : request;
+
+    string[string] urlParams = [
+        "method" : method,
+    ];
+    static if (!is(P == typeof(null)))
```
Comment (ZombineDev) [inline source/dlangbot/bugzilla.d]: Hmm, not sure what this line is supposed to mean, but to me this looks like `version (none)`.

File: source/dlangbot/bugzilla.d [hunk 3]
```diff
@@ -79,3 +81,43 @@ Issue[] getDescriptions(R)(R issueRefs)
...
+}
+
+Json authenticatedApiCall(string method, Json[string] params)
+{
+    params["Bugzilla_login"] = bugzillaLogin;
+    params["Bugzilla_password"] = bugzillaPassword;
+    return apiCall(method, params);
+}
+
+/// Close these bug IDs as FIXED and leave a comment.
```
Comment (ZombineDev) [inline source/dlangbot/bugzilla.d]: Nit: order higher-level / public functions before lower level / private ones - `closeIssues` first, and then `authenticatedApiCall`, followed by `apiCall`.

File: source/dlangbot/bugzilla.d [hunk 4]
```diff
@@ -79,3 +81,43 @@ Issue[] getDescriptions(R)(R issueRefs)
...
+void closeIssues(int[] bugIDs, string comment)
```
Comment (ZombineDev) [inline source/dlangbot/bugzilla.d]:
```suggestion
void closeIssues(uint[] bugIDs, string comment)
```

File: source/dlangbot/bugzilla.d [hunk 5]
```diff
@@ -79,3 +81,43 @@ Issue[] getDescriptions(R)(R issueRefs)
...
+{
+    authenticatedApiCall("Bug.update", [
+        "ids" : bugIDs.map!(id => Json(id)).array.Json,
+        "status" : "RESOLVED".Json,
+        "resolution" : "FIXED".Json,
+        "comment" : [
+            "body" : comment.Json,
+        ].Json,
```
[the remark this answers is not in the document]
Comment (CyberShadow) [reply source/dlangbot/bugzilla.d]: No, the `Json` constructor for JSON arrays wants an actual `Json[]`.

There is the deserialization stuff in the same module, but that goes through a different code path (introspection etc.).

File: source/dlangbot/bugzilla.d [hunk 6]
```diff
@@ -79,3 +81,43 @@ Issue[] getDescriptions(R)(R issueRefs)
         .sort!((a, b) => a.id < b.id)
         .release;
 }
+
+shared string bugzillaLogin, bugzillaPassword;
```
[the remark this answers is not in the document]
Comment (CyberShadow) [reply source/dlangbot/bugzilla.d]: Yep, but also, these variables are set to different values in the tests.

File: source/dlangbot/bugzilla.d (hunk 2 above)
Comment (CyberShadow) [reply source/dlangbot/bugzilla.d]: Oops, that's a left-over from an older implementation

File: source/dlangbot/bugzilla.d (hunk 4 above)
Comment (CyberShadow) [reply source/dlangbot/bugzilla.d]: Issue numbers are `int` elsewhere.

**Comment by CyberShadow** [comment]
I can't figure out how to run the damn tests.

Problems I ran into:
- Compilation errors due to newer DMDs no longer being able to compile Vibe.d
- Linking errors due to OpenSSL 1.1
- Compilation errors due to `_expectedStatusCode` being declared in `version(unittest)` but accessed outside `version(unittest)`

I got as far as:

```console
$ dver -d 2.078.2 env DFLAGS="-L/usr/lib/libssl.so.1.0.0 -L/usr/lib/libcrypto.so.1.0.0" dub test --compiler=dmd -- --single --trace
```

which seems to match the command / environment exactly as in Travis CI, but I get:

```
ut.d(5,5): Error: only one main allowed. Previously found main at source/dlangbot/app.d(380,11)
```

**Comment by wilzbach** [comment]
> * Linking errors due to OpenSSL 1.1

`--override-config vibe-d:tls/openssl-1.1` (though the upcoming 0.8.5 will have auto-detection)

> * Compilation errors due to newer DMDs no longer being able to compile Vibe.d

Compiles fine with 2.084.0 here. What error do you get?

> * Compilation errors due to `_expectedStatusCode` being declared in `version(unittest)` but accessed outside `version(unittest)`

I never saw that one before.

> I can't figure out how to run the damn tests.

The command that I posted still works fine for me:

```
> dmd --version

DMD64 D Compiler v2.084.0

> dub test --override-config vibe-d:tls/openssl-1.1 -- -s
The sub configuration directive "vibe-d:core" -> "vibe-core" references a package that is not specified as a dependency and will have no effect.
Running custom 'unittest' configuration.
Performing "unittest" build using /usr/bin/dmd for x86_64.
unit-threaded 0.7.46: target for configuration "library" is up to date.
taggedalgebraic 0.10.11: target for configuration "library" is up to date.
eventcore 0.8.35: target for configuration "epoll" is up to date.
stdx-allocator 2.77.2: target for configuration "library" is up to date.
vibe-core 1.4.1: target for configuration "epoll" is up to date.
vibe-d:utils 0.8.4: target for configuration "library" is up to date.
vibe-d:data 0.8.4: target for configuration "library" is up to date.
mir-linux-kernel 1.0.0: target for configuration "library" is up to date.
vibe-d:crypto 0.8.4: target for configuration "library" is up to date.
diet-ng 1.5.0: target for configuration "library" is up to date.
vibe-d:stream 0.8.4: target for configuration "library" is up to date.
vibe-d:textfilter 0.8.4: target for configuration "library" is up to date.
vibe-d:inet 0.8.4: target for configuration "library" is up to date.
vibe-d:tls 0.8.4: target for configuration "openssl-1.1" is up to date.
vibe-d:http 0.8.4: target for configuration "library" is up to date.
vibe-d:mail 0.8.4: target for configuration "library" is up to date.
vibe-d:mongodb 0.8.4: target for configuration "library" is up to date.
vibe-d:redis 0.8.4: target for configuration "library" is up to date.
vibe-d:web 0.8.4: target for configuration "library" is up to date.
vibe-d 0.8.4: target for configuration "vibe-core" is up to date.
dlang-bot ~master: building configuration "unittest"...
Running pre-build commands...
Building package unit-threaded in /home/seb/.dub/packages/unit-threaded-0.7.46/unit-threaded/
Performing "$DFLAGS" build using /usr/bin/dmd for x86_64.
unit-threaded 0.7.46: target for configuration "gen_ut_main" is up to date.
To force a rebuild of up-to-date targets, run again with --force.
Running ../../.dub/packages/unit-threaded-0.7.46/unit-threaded/gen_ut_main -f ut.d
Compiling Diet HTML template index.dt...
Warning: testCronDaily passes the criteria for a value-parameterized test function but doesn't have the appropriate value UDAs.
         Consider changing its name or annotating it with @DontTest
Linking...
To force a rebuild of up-to-date targets, run again with --force.
Running ./dlang-bot -s
[main(----) INF] Listening for requests on http://0.0.0.0:36957/
[main(----) INF] Listening for requests on http://0.0.0.0:43089/
[main(----) INF] Starting test in test/buildkite.d:14 with payload: ./data/hooks/buildkite/ping.json
buildkite.answers-ping:
buildkite.does-not-reap-running-servers:
...
```

**Comment by wilzbach** [comment]
I would recommend to just bump the release compiler as DMD latest passed fine on Travis:

https://travis-ci.org/dlang/dlang-bot/builds/490048932

It's the `.d-compiler` (Heroku) and of course the fixed job in `.travis.yml`.

**Comment by wilzbach** [comment]
BTW an interesting observation I just made is that if I `dub upgrade` (on the master branch), I get the same error as the one failing Travis job (exit with `-6`).
A lot of things changed with a `dub upgrade` though:

```diff
diff --git a/dub.selections.json b/dub.selections.json
index fdde7ad..35fc7ab 100644
--- a/dub.selections.json
+++ b/dub.selections.json
@@ -4,16 +4,16 @@
 		"botan": "1.12.10",
 		"botan-math": "1.0.3",
 		"diet-ng": "1.5.0",
-		"eventcore": "0.8.35",
+		"eventcore": "0.8.40",
 		"libasync": "0.8.3",
 		"libevent": "2.0.2+2.0.16",
-		"memutils": "0.4.11",
-		"mir-linux-kernel": "1.0.0",
+		"memutils": "0.4.13",
+		"mir-linux-kernel": "1.0.1",
 		"openssl": "1.1.6+1.0.1g",
-		"stdx-allocator": "2.77.2",
-		"taggedalgebraic": "0.10.11",
-		"unit-threaded": "0.7.46",
-		"vibe-core": "1.4.1",
+		"stdx-allocator": "2.77.5",
+		"taggedalgebraic": "0.10.12",
+		"unit-threaded": "0.7.55",
+		"vibe-core": "1.6.0",
 		"vibe-d": "0.8.4"
 	}
 }
```

**Comment by CyberShadow** [comment]
> Compiles fine with 2.084.0 here. What error do you get?

```
../../../../.dub/packages/vibe-d-0.8.2/vibe-d/utils/vibe/internal/memory_legacy.d(187,37): Error: constructor vibe.utils.hashmap.HashMap!(void*, ulong, DefaultHashMapTraits!(void*)).HashMap.this(IAllocator allocator) is not callable using argument types (RCIAllocator)
../../../../.dub/packages/vibe-d-0.8.2/vibe-d/utils/vibe/internal/memory_legacy.d(187,37):        cannot pass argument (*function () => allocatorObject(instance))() of type RCIAllocator to parameter IAllocator allocator
../../../../.dub/packages/vibe-d-0.8.2/vibe-d/utils/vibe/internal/utilallocator.d(14,85): Error: cannot implicitly convert expression (*function () => allocatorObject(instance))() of type RCIAllocator to std.experimental.allocator.IAllocator
```

I already tried deleting `~/.dub`. Anything else I could try?

> dub test --override-config vibe-d:tls/openssl-1.1 -- -s

That fails for me with the above errors (with DMD 2.084.0).

> BTW an interesting observation I just made is that if I `dub upgrade` (on the master branch), I get the same error as the one failing Travis job (exit with `-6`).

The errors on Travis look like they're due to unstable AA iteration order.

I tried fixing this yesterday. This sucks. Vibe's JSON module uses a `Json[string]` AA to represent objects, and even its serialization code uses `Json` as an intermediary format.

Ways of fixing this:
- Hard-code all possible strings resulting from the AA iteration order in the compiler versions and architectures we support.
- Uglify the code added here by completely avoiding string AAs or JSON objects. JSON object stringification would then need to be done manually.
- Add a dependency on a JSON library which doesn't have this problem (such as `ae.utils.json`).

**Comment by wilzbach** [comment]
Where do you get 0.8.2 from? That's a release from more than a year ago!
Maybe you have it globally installed somewhere?
Try looking at `dub test ... -v`

**Comment by CyberShadow** [comment]
> @CyberShadow CyberShadow force-pushed the CyberShadow:pull-20190206-224050 branch from 9e4d6b7 to d895eb1 21 hours ago

No idea how this happened, but that force-push rolled back dub.selections.json to a much older version. I don't understand how I didn't notice the staged changes, or where the old version even came from. Maybe my editor ran a Dub command during the rebase, which then updated the file...

**Comment by wilzbach** [comment]
> The errors on Travis look like they're due to unstable AA iteration order.

Hmm it's fully reproducible and happens every time (at least after the dub upgrade). I would rather guess that it has sth. to do with a change in vibe-core, but as mentioned we don't have to look into this regression (?) for now. It still works with the current set of dependencies in the `dub.selection.json`.

**Comment by CyberShadow** [comment]
I was still getting the `ut.d` error after rebuilding with the fixed `dub.selections.json`, but that went away after a `git clean -fdx`. I can finally run the tests now!

> Hmm it's fully reproducible and happens every time (at least after the dub upgrade).

Just to we're on the same page, I was referring to this problem:

https://travis-ci.org/dlang/dlang-bot/jobs/490048935#L558

By "unstable" I didn't mean differing from run to run, but rather, changing between versions/architectures.

File: test/utils.d [hunk 7]
```diff
@@ -109,17 +111,19 @@ auto payloadServer(scope HTTPServerRequest req, scope HTTPServerResponse res)
...
+        logError("Request for unexpected URL received: '%s'", req.requestURL);
+        assert(0);
     }
 
     res.statusCode = expectation.respStatusCode;
     // set failure status code exception to suppress false errors
-    import dlangbot.utils : _expectedStatusCode;
-    if (expectation.respStatusCode / 100 != 2)
-        _expectedStatusCode = expectation.respStatusCode;
+    version (unittest)
```
Comment (wilzbach) [inline test/utils.d]: Do you still need this `version(unittest)` now that you fixed your dub build?
It's in the `test` folder (which for dub implies it's a `unittest` build).

**Comment by wilzbach** [comment]
> I was still getting the `ut.d` error after rebuilding with the fixed `dub.selections.json`, but that went away after a `git clean -fdx`. I can finally run the tests now!

Awesome.

> Just to we're on the same page, I was referring to this problem:

Ah I see (the error you get when upgrading the dub dependencies is different).
With 2.082 the hashing function for the dictionary was changed [1]. So that's why imho the easiest way forward here is to bump the host compiler to latest. Alternatively, instead of `params.Json.toString()`, you could sort the keys and build the `params` string by looping over the ordered keys?

[1] https://dlang.org/changelog/2.082.0.html#typeinfo-gethash

File: source/dlangbot/bugzilla.d [hunk 8]
```diff
@@ -79,3 +81,42 @@ Issue[] getDescriptions(R)(R issueRefs)
...
+Json apiCall(string method, Json[string] params)
+{
+    import vibe.stream.operations : readAllUTF8;
+    import dlangbot.utils : request;
+
+    string[string] urlParams = [
+        "method" : method,
+    ];
+    if (params)
+        urlParams["params"] = params.Json.toString();
```
Comment (wilzbach) [inline source/dlangbot/bugzilla.d]: This `toString()` is where the unstable sorting comes from.

Comment (CyberShadow) [reply source/dlangbot/bugzilla.d]: Also `urlEncode`, but that's easy to avoid.

**Comment by CyberShadow** [comment]
> `--override-config vibe-d:tls/openssl-1.1`

That links successfully and runs the test suite, but the built binary can't actually make HTTPS requests. Fails with:

```
object.Exception@../../../../.dub/packages/vibe-d-0.8.4/vibe-d/tls/vibe/stream/openssl.d(381): Connecting TLS tunnel:  (0)
```

> `DFLAGS="-L/usr/lib/libssl.so.1.0.0 -L/usr/lib/libcrypto.so.1.0.0"`

Adding these options works with `rdmd` / `rund` / `dmd -i`, but the above doesn't work with Dub (linking fails as if the var wasn't set). Possibly because Dub links as a separate, explicit step, or because it doesn't support `DFLAGS`.

In the end the only thing that worked for me was to add `lflags "/usr/lib/libssl.so.1.0.0" "/usr/lib/libcrypto.so.1.0.0"` to `dub.sdl`. I couldn't find a way to specify those on the command line in Dub's `--help` output.

**Comment by CyberShadow** [comment]
> What if the issue is already closed? Need to check if the Bugzilla API errors out. We should post a comment regardless.

It remains closed and the comment gets posted, no problem.

I tested the issue closing against live Bugzilla and made fixes appropriately. With that, I think this is good to go / final review round.

**Comment by wilzbach** [comment]
> but the built binary can't actually make HTTPS requests.

Really weird. Have you done anything specific in the crypto/security area to your Arch setup?

> Possibly because Dub links as a separate, explicit step, 

Yup.

> In the end the only thing that worked for me was to add lflags "/usr/lib/libssl.so.1.0.0" "/usr/lib/libcrypto.so.1.0.0" to dub.sdl. 

Isn't this dub's current default? (i.e. what it tries to link to by default)

**Comment by CyberShadow** [comment]
> Have you done anything specific in the crypto/security area to your Arch setup?

No. I do have https://www.archlinux.org/packages/core/x86_64/openssl-1.0/ installed (it's a Steam dependency).

> Isn't this dub's current default? (i.e. what it tries to link to by default)

No, it just uses `-lssl -lcrypto` by default.

**Comment by CyberShadow** [comment]
Oops, thanks for reminding. I can no longer reproduce that problem, so I'm dropping that commit.

**Comment by wilzbach** [comment]
Let's keep an eye on the bot for the next day to ensure this works as intended.

**Comment by wilzbach** [comment]
The first real-world test passed:

https://issues.dlang.org/show_bug.cgi?id=19676

**Comment by CyberShadow** [comment]
> Thanks so much for this and still finishing it despite all the annoyances with Vibe.d.

Thanks for holding my hand through the ordeal. I doubt I would have made it without help.

I ran into one more problem: I was not getting any information at all except the error message (which was often just "Assertion failure") from thrown / uncaught exceptions. https://github.com/dlang/dlang-bot/pull/225/commits/eed1a1a3ed6d29e03febedec3550c9d4f5252214 helped for some cases. For stuff thrown in `postGitHubHook`, wrapping it in `try` ... `catch (Throwable e){ logError(e.toString()); throw e; }` also allowed me to see the error. However, this didn't let me see what caused errors thrown from other places. The logs emitted by unit-threaded also didn't make any sense - the program would just crash with `-6` (which I think is SIGABRT). In the end, my solution was to bypass Dub / Unit-Threaded and build individual test files "manually" with `dmd -i`, which finally allowed me to see the stack traces.
